src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java

```diff
@@ -0,0 +1,207 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.processor.data.entity;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.spongepowered.common.data.util.DataUtil.getData;
+
+import com.google.common.base.Optional;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.DataHolder;
+import org.spongepowered.api.data.DataTransactionBuilder;
+import org.spongepowered.api.data.DataTransactionResult;
+import org.spongepowered.api.data.DataView;
+import org.spongepowered.api.data.key.Key;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.immutable.entity.ImmutablePassengerData;
+import org.spongepowered.api.data.manipulator.mutable.entity.PassengerData;
+import org.spongepowered.api.data.merge.MergeFunction;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.api.service.persistence.InvalidDataException;
+import org.spongepowered.common.data.DataProcessor;
+import org.spongepowered.common.data.manipulator.immutable.entity.ImmutableSpongePassengerData;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongePassengerData;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+
+public class PassengerDataProcessor implements DataProcessor<PassengerData, ImmutablePassengerData> {
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof net.minecraft.entity.Entity;
+    }
+
+    @SuppressWarnings("unused")
+    @Override
+    public Optional<PassengerData> from(DataHolder dataHolder) {
+        if (dataHolder instanceof net.minecraft.entity.Entity) {
+            final SpongePassengerData passengerData = new SpongePassengerData();
+            final Entity vehicle = (Entity) ((net.minecraft.entity.Entity) dataHolder).ridingEntity;
+
+            net.minecraft.entity.Entity currentEntity = (net.minecraft.entity.Entity) vehicle;
+
+            while (currentEntity != null && currentEntity.isRiding())
+            {
+                currentEntity = currentEntity.ridingEntity;
+            }
+
+            final Entity baseVehicle = (Entity) currentEntity;
+
+            return Optional.<PassengerData>of(passengerData.setVehicle(vehicle).setBaseVehicle(baseVehicle));
+        } else {
+            return Optional.absent();
+        }
+
+    }
+
+    @Override
+    public Optional<PassengerData> fill(DataHolder dataHolder, PassengerData manipulator) {
+        if (dataHolder instanceof net.minecraft.entity.Entity) {
+
+            manipulator.set(Keys.VEHICLE, (Entity) ((net.minecraft.entity.Entity) dataHolder).ridingEntity);
```

Comment:
You're not checking whether the ridingEntity is null, which you should do first, if the riding entity is null, you shouldn't be setting it on the manipulator.